### PR TITLE
DogFood: Use OpenDevin to review PR

### LINKED ISSUE #1178: DogFood: Use OpenDevin to review PRs
**What problem or use case are you trying to solve?**

It would be nice if we could let OpenDevin review PRs in OpenDevin repo. Inspired by https://github.com/OpenDevin/OpenDevin/issues/1067.

**Describe the UX of the solution you'd like**

No UX is needed. We need a GitHub workflow file.

**Do you have thoughts on the technical implementation?**

Reviewing PRs sounds easy, but in some sense, it's harder than creating a PR. To review a PR, the agent essentially needs to:
1) Understand the patch
2) Fix/Modify the patch
3) Verify the patch works
4) Suggest the updated patch

Of course, at the moment, we can simply let the agent review the patch and post a single review comment. Eventually it shall be able to submit review at appropriate lines, and/or even `Add a suggestion` (one-liner code changes that can be directly applied on GitHub GUI) directly.

**Additional context**

I am working on this. Will come up with a preliminary version soon.

### PR DESCRIPTION
**Pull request by li-boxuan**
This PR introduces a GitHub Action that aims to review PRs automatically, if a label `dogfood-this`
is present.

TL;DR: it works, but due to the limitations & potential bugs of OpenDevin, the review bot is not very useful at the moment. Maybe our PRs are already of high quality so it cannot detect anything wrong (that is obvious enough)?

## Limitations:
1) Experiments show we have to include the patch content in the task itself. If we tell
OpenDevin to read the patch file and review it, it *sometimes* complains the file does
not exist. This doesn't seem to be a problem of LLM per se, but a bug of our
sandbox. For example, I've witnessed multiple times that OpenDevin complains that
`ls` returns nothing, but at other times, it is able to use `cat` to read the file content.
2) Although instructed, the agent doesn't attempt to understand the patch within the
project context.
3) Review is posted as a single comment, rather at specific lines.

### Case 1: https://github.com/li-boxuan/OpenDevin/pull/1

> The pull request is fixing the main html overflow by adding 'overflow: hidden' to the tag. This fix will prevent the main HTML content from overflowing and provide a better user experience.

### Case 2: https://github.com/li-boxuan/OpenDevin/pull/2

> I have completed a detailed review of the pull request and provided my comments and suggestions in the review-2.txt file. It seems like the pull request demonstrates good changes to the documentation, including fixing broken links and providing relevant information. Additionally, the addition of specific LLM guides enhances the user experience. However, there is a minor typo on line 144 of the README.md file that should be corrected for clarity. Overall, the changes are positive and helpful to the project.

Comment: I don't see any typo around line 144 :(

Note: the log shows OpenDevin hits an exception:

> 07:17:43 - opendevin:ERROR: agent_controller.py:178 - AgentFinishAction.__init__() got an unexpected keyword argument 'final_thought'
07:17:43 - OBSERVATION
AgentFinishAction.__init__() got an unexpected keyword argument 'final_thought'

Seems we could have better error handling here.

### Case 3: https://github.com/li-boxuan/OpenDevin/pull/3

> The pull request looks good overall. The changes made in the ghcr.yml file seem to fix the workflow error in the fork repo. Additionally, the changes in the containers seem to address the Docker registry and organization appropriately. It may be beneficial to clarify the purpose of the changes in the pull request description for better context. Overall, the changes look good and seem to address the intended issues. Nice work!

### Case 4: https://github.com/li-boxuan/OpenDevin/pull/4

This is an interesting one. I had to cancel the workflow because it seems the agent was not able to finish by itself. This PR's diff contains a lot of prompts, which apparently made agent confused. In step 0, the agent attempted to:

> 07:06:46 - ACTION
BrowseURLAction(url='https://github.com/pull-request-review-4', action=<ActionType.BROWSE: 'browse'>)

which makes no sense.

Closes #1178

### CODE REVIEW

**Comment by enyst** [comment]
Fascinating! <3

A small curiosity:

> Note: the log shows OpenDevin hits an exception:
> 07:17:43 - opendevin:ERROR: agent_controller.py:178 - AgentFinishAction.init() got an unexpected keyword argument 'final_thought'
07:17:43 - OBSERVATION
AgentFinishAction.init() got an unexpected keyword argument 'final_thought'
Seems we could have better error handling here

If I understand correctly, the model was GPT-3.5?

File: .github/workflows/review-pr.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  pull_request:
+    types: [synchronize, labeled]
+
+permissions:
+  contents: write
+  pull-requests: write
+
+jobs:
+  open-devin:
+    if: contains(github.event.pull_request.labels.*.name, 'dogfood-this')
```
Comment (foragerr) [inline .github/workflows/review-pr.yml:13]: Consider using a different label, perhaps `review-this`
so it's distinct from the solve-issue workflow?

( Yup, the other workflow could use a better label too :D )

Comment (li-boxuan) [reply .github/workflows/review-pr.yml:13]: Agreed. `dogfood` is not wrong but isn't concise enough.

**Comment by foragerr** [comment]
Agreed, the review quality isn't great. It was the same issue with the other workflow, it could only do very trivial tasks. But there's work in flight to improve agent quality over time.

I wonder if this one will benefit from prompt engineering. See: https://github.com/coderabbitai/ai-pr-reviewer/blob/main/src/prompts.ts#L3

File: .github/workflows/review-pr.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        echo "Your coworker wants to apply a pull request to this project. Read and review it, and write your concise comments and suggestions to review-${{ github.event.pull_request.number }}.txt" > task.txt
+        echo "TITLE:" >> task.txt
+        echo "${{ github.event.pull_request.title }}" >> task.txt
+        echo "" >> task.txt
+        echo "BODY:" >> task.txt
+        echo "${{ github.event.pull_request.body }}" >> task.txt
+        echo "" >> task.txt
+        echo "The diff is:" >> task.txt
+        cat ${{ github.event.pull_request.number }}.diff >> task.txt
+        rm ${{ github.event.pull_request.number }}.diff
```
[the remark this answers is not in the document]
Comment (li-boxuan) [reply .github/workflows/review-pr.yml:46]: > Does this diff contain line numbers?

Yes, but there's only a line number for the beginning of each section. Example: https://patch-diff.githubusercontent.com/raw/OpenDevin/OpenDevin/pull/1223.patch

[the remark this answers is not in the document]
Comment (li-boxuan) [reply .github/workflows/review-pr.yml:46]: > I wonder if the agent can be made to produce a SARIF file.

I failed to understand how SARIF is used, but seems it's out of context here? You might wanna create a new issue?

**Comment by li-boxuan** [comment]
> Agreed, the review quality isn't great. It was the same issue with the other workflow, it could only do very trivial tasks

Yep. Actually I think currently the pr review bot does better than the pr solver bot, because it's much easier to give a mediocre review comment than a mediocre (but working) solution.

**Comment by li-boxuan** [comment]
> I wonder if this one will benefit from prompt engineering

Yeah I think so! But that doesn't seem to align with the goal of OpenDevin (this is just my personal understanding of OpenDevin's mission, though): we shouldn't need sophisticated prompt engineering when we use OpenDevin.

Maybe a long-term solution would be to create a new agent (with those prompt engineering)? An agent which could review the other SWE agent's code and let them fight with each other.

**Comment by li-boxuan** [comment]
I feel like https://github.com/OpenDevin/OpenDevin/pull/1238 is the way to go. Create a mini reviewer agent, and the main/meta agent should automatically be able to hand over the work to reviewer agent.

**Review by rbren — APPROVED** [review]
This is neat! Let's give it a shot

File: .github/workflows/review-pr.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (foragerr) [reply .github/workflows/review-pr.yml:46]: Producing and uploading a SARIF file is how you'd do comments on specific lines, like you mentioned in your # 3 
> Review is posted as a single comment, rather at specific lines.

[the remark this answers is not in the document]
Comment (li-boxuan) [reply .github/workflows/review-pr.yml:46]: Aha, I see what you are saying. That sounds very interesting! Seems a way to go since `gh` cli currently doesn't support fine-grained review: https://github.com/cli/cli/issues/359

This https://dxuuu.xyz/prr.html may also be worth trying.

[the remark this answers is not in the document]
Comment (li-boxuan) [reply .github/workflows/review-pr.yml:46]: Created https://github.com/OpenDevin/OpenDevin/issues/1254
